core/message_pool.go

```diff
@@ -38,21 +63,22 @@ type BlockTimer interface {
 type MessagePool struct {
 	lk sync.RWMutex
 
-	timer   BlockTimer
-	pending map[cid.Cid]*timedmessage // all pending messages
+	api           MessagePoolAPI
+	pending       map[cid.Cid]*timedmessage // all pending messages
+	addressNonces map[addressNonce]bool
 }
 
 // Add adds a message to the pool.
-func (pool *MessagePool) Add(msg *types.SignedMessage) (cid.Cid, error) {
-	blockTime, err := pool.timer.BlockHeight()
+func (pool *MessagePool) Add(ctx context.Context, msg *types.SignedMessage) (cid.Cid, error) {
+	blockTime, err := pool.api.BlockHeight()
 	if err != nil {
 		return cid.Undef, err
 	}
 
-	return pool.addTimedMessage(&timedmessage{message: msg, addedAt: blockTime})
+	return pool.addTimedMessage(ctx, &timedmessage{message: msg, addedAt: blockTime})
 }
 
-func (pool *MessagePool) addTimedMessage(msg *timedmessage) (cid.Cid, error) {
+func (pool *MessagePool) addTimedMessage(ctx context.Context, msg *timedmessage) (cid.Cid, error) {
```

Comment:
prolly worth a comment that if this thing returns error it could be because of something Bad eg can't compute a cid or it could be because the message is not valid which is expected. if you wanted to distinguish we could move validation out of addtimedmessage at the cost of a little bit of danger that someone adds unvalidated messages.

Reply:
done